util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,398 @@
+            fprintf(stderr, "Error: packet dimensionality doesn't match previous packets of same Func. Aborting.\n");
+            exit(-1);
+        }
+
+        for (int lane = 0; lane < lanes; lane++) {
+            for (int i = 0; i < real_dims; i++) {
+                if (p->coordinates()[lanes*i+lane] < min_coords[i]) {
+                    min_coords[i] = p->coordinates()[lanes*i+lane];
+                }
+                if (p->coordinates()[lanes*i+lane] > max_coords[i]) {
+                    max_coords[i] = p->coordinates()[lanes*i+lane];
+                }
+            }
+        }
+    }
+
+    void allocate() {
+        std::vector<int> extents;
+        for (int i = 0; i < dimensions; i++) {
+            extents.push_back(max_coords[i] - min_coords[i] + 1);
+        }
+        values = Buffer<>(type, extents);
+        if (values.data() == nullptr) {
+            fprintf(stderr, "Memory allocation failure. Aborting.\n");
+            exit(-1);
+        }
+    }
+
+    void add(Packet *p) {
+        halide_type_t scalar_type = p->type;
+        scalar_type.lanes = 1;
+        if (scalar_type == halide_type_of<float>()) {
+            add_typed<float>(p);
+        } else if (scalar_type == halide_type_of<double>()) {
+            add_typed<double>(p);
+        } else if (scalar_type == halide_type_of<uint8_t>()) {
+            add_typed<uint8_t>(p);
+        } else if (scalar_type == halide_type_of<uint16_t>()) {
+            add_typed<uint16_t>(p);
+        } else if (scalar_type == halide_type_of<uint32_t>()) {
+            add_typed<uint32_t>(p);
+        } else if (scalar_type == halide_type_of<uint64_t>()) {
+            add_typed<uint64_t>(p);
+        } else if (scalar_type == halide_type_of<int8_t>()) {
+            add_typed<int8_t>(p);
+        } else if (scalar_type == halide_type_of<int16_t>()) {
+            add_typed<int16_t>(p);
+        } else if (scalar_type == halide_type_of<int32_t>()) {
+            add_typed<int32_t>(p);
+        } else if (scalar_type == halide_type_of<int64_t>()) {
+            add_typed<int64_t>(p);
+        } else if (scalar_type == halide_type_of<bool>()) {
+            add_typed<bool>(p);
+        } else {
+            printf("Packet with unknown type\n");
+            exit(-1);
+        }
+    }
+
+    template<typename T>
+    void add_typed(Packet *p) {
+        Buffer<T> &buf = values.as<T>();
+        int lanes = p->type.lanes;
+
+        if (!allocated()) {
+            fprintf(stderr, "Packet storage not allocated. Aborting.\n");
+            exit(-1);
+        }
+
+        if (p->dimensions < 0) {
+            fprintf(stderr, "Negative dimensionality found. Aborting.\n");
+            exit(-1);
+        }
+
+        for (int lane = 0; lane < lanes; lane++) {
+            int coord[16];
+            for (int i = 0; i < dimensions; i++) {
+                coord[i] = p->coordinates()[lanes * i + lane] - min_coords[i];
+            }
+            buf(coord) = p->get_value_as<T>(lane);
```

Comment:
does it matter than some uninitialized entries in coord[] are passed to buf?

Reply:
No, it wants an int array of size dimensions.